File: test_python/memory_actions_test.py
import sys
import unittest
from pathlib import Path

SCRIPT_DIR = Path(__file__).parents[1] / "skills" / "tokensmind-agent-network" / "scripts"
sys.path.insert(0, str(SCRIPT_DIR))

from agent_network_runtime import ActionExecutor  # noqa: E402


class MemoryWorkflow:
    def __init__(self):
        self.value = None

    def read(self):
        return self.value

    def write(self, value):
        self.value = value

    def remove(self):
        self.value = None
        return True


class FakeApi:
    def __init__(self, responses):
        self.responses = list(responses)
        self.calls = []

    def request(self, method, path, body=None, *, key=None):
        self.calls.append({"method": method, "path": path, "body": body, "key": key})
        return self.responses.pop(0)


def explicit_memory():
    return {
        "agentId": "untrusted-agent",
        "clientMemoryId": "memory-client-1",
        "confidence": 1,
        "contextText": "Frontend implementation tasks",
        "kind": "matching_preference",
        "polarity": "positive",
        "scope": "contextual",
        "sourceType": "user_explicit",
        "statement": "Prefer Agents that deliver working code quickly.",
    }


class MemoryActionsTest(unittest.TestCase):
    OWNED_AGENT = {"id": "owned/agent", "name": "Memory Agent"}

    def execute(self, responses, operation, data):
        api = FakeApi(responses)
        result = ActionExecutor(api, MemoryWorkflow()).execute({"operation": operation, "input": data})
        return api, result

    def test_get_settings_uses_owned_agent(self):
        api, result = self.execute(
            [[self.OWNED_AGENT], {"collectionEnabled": True, "matchingEnabled": True}],
            "get_memory_settings", {"agentId": "untrusted-agent"},
        )
        self.assertEqual(result["status"], "completed")
        self.assertEqual(api.calls[1]["path"], "/agent-network-api/agents/owned%2Fagent/memory-settings")

    def test_propose_memory_has_node_parity(self):
        api, result = self.execute(
            [[self.OWNED_AGENT], {"collectionEnabled": True}, {"id": "memory-1"}],
            "propose_memory", explicit_memory(),
        )
        self.assertEqual(result["status"], "completed")
        self.assertEqual(api.calls[2]["path"], "/agent-network-api/agents/owned%2Fagent/memories")
        self.assertTrue(api.calls[2]["key"].endswith(":memory:propose"))
        self.assertNotIn("agentId", api.calls[2]["body"])

    def test_disabled_collection_stops_before_write(self):
        api, result = self.execute(
            [[self.OWNED_AGENT], {"collectionEnabled": False}],
            "propose_memory", explicit_memory(),
        )
        self.assertEqual(result["status"], "failed")
        self.assertEqual(result["code"], "MEMORY_COLLECTION_DISABLED")
        self.assertEqual(len(api.calls), 2)

    def test_inferred_confidence_is_validated_before_requests(self):
        data = {**explicit_memory(), "sourceType": "model_inferred", "confidence": 0.9}
        api, result = self.execute([], "propose_memory", data)
        self.assertEqual(result["status"], "input_required")
        self.assertEqual(result["fields"], ["confidence"])
        self.assertEqual(api.calls, [])

    def test_list_memories_encodes_filters(self):
        api, result = self.execute(
            [[self.OWNED_AGENT], {"items": [], "total": 0}],
            "list_memories", {"limit": 25, "offset": 5, "status": "pending_review"},
        )
        self.assertEqual(result["status"], "completed")
        self.assertEqual(
            api.calls[1]["path"],
            "/agent-network-api/agents/owned%2Fagent/memories?limit=25&offset=5&status=pending_review",
        )

    def test_delete_memory_uses_owned_agent_and_stable_key(self):
        api, result = self.execute(
            [[self.OWNED_AGENT], {"deleted": True}],
            "delete_memory", {"agentId": "untrusted-agent", "memoryId": "memory/id"},
        )
        self.assertEqual(result["status"], "completed")
        self.assertEqual(api.calls[1]["path"], "/agent-network-api/agents/owned%2Fagent/memories/memory%2Fid")
        self.assertTrue(api.calls[1]["key"].endswith(":memory:delete"))


if __name__ == "__main__":
    unittest.main()

File: test/memory-actions.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { readFile } from 'node:fs/promises';
import { createActionExecutor } from '../src/action-executor.js';

const OWNED_AGENT = Object.freeze({ id: 'owned/agent', name: 'Memory Agent' });

function executorFor(responses) {
  const calls = [];
  let workflow = null;
  return {
    calls,
    executor: createActionExecutor({
      api: {
        async request(request) {
          calls.push(request);
          const response = responses.shift();
          if (response instanceof Error) throw response;
          return response;
        },
      },
      workflowStore: {
        async read() { return workflow; },
        async write(_name, value) { workflow = value; },
        async remove() { workflow = null; return true; },
      },
      randomId: () => 'workflow-1',
    }),
  };
}

function explicitMemory() {
  return {
    agentId: 'untrusted-agent',
    clientMemoryId: 'memory-client-1',
    confidence: 1,
    contextText: 'Frontend implementation tasks',
    kind: 'matching_preference',
    polarity: 'positive',
    scope: 'contextual',
    sourceType: 'user_explicit',
    statement: 'Prefer Agents that deliver working code quickly.',
  };
}

test('get_memory_settings resolves the owned Agent before reading settings', async () => {
  const found = executorFor([[OWNED_AGENT], { collectionEnabled: true, matchingEnabled: true }]);
  const result = await found.executor.execute({
    operation: 'get_memory_settings', input: { agentId: 'untrusted-agent' },
  });

  assert.equal(result.status, 'completed');
  assert.deepEqual(found.calls.map(({ method, path }) => `${method} ${path}`), [
    'GET /agent-network-api/agents?mine=1',
    'GET /agent-network-api/agents/owned%2Fagent/memory-settings',
  ]);
});

test('propose_memory checks collection and sends the minimal normalized body', async () => {
  const found = executorFor([[OWNED_AGENT], { collectionEnabled: true }, { id: 'memory-1' }]);
  const result = await found.executor.execute({ operation: 'propose_memory', input: explicitMemory() });

  assert.equal(result.status, 'completed');
  assert.equal(found.calls[2].path, '/agent-network-api/agents/owned%2Fagent/memories');
  assert.equal(found.calls[2].idempotencyKey, 'workflow-1:memory:propose');
  assert.equal(Object.hasOwn(found.calls[2].body, 'agentId'), false);
  assert.deepEqual(found.calls[2].body, {
    clientMemoryId: 'memory-client-1',
    confidence: 1,
    contextText: 'Frontend implementation tasks',
    conversationId: null,
    expiresAt: null,
    kind: 'matching_preference',
    polarity: 'positive',
    requirementId: null,
    scope: 'contextual',
    sourceType: 'user_explicit',
    statement: 'Prefer Agents that deliver working code quickly.',
    targetAgentId: null,
  });
});

test('propose_memory fails explicitly when collection is disabled', async () => {
  const found = executorFor([[OWNED_AGENT], { collectionEnabled: false }]);
  const result = await found.executor.execute({ operation: 'propose_memory', input: explicitMemory() });

  assert.equal(result.status, 'failed');
  assert.equal(result.code, 'MEMORY_COLLECTION_DISABLED');
  assert.equal(found.calls.length, 2);
});

test('inferred memory is capped by validation before any request', async () => {
  const found = executorFor([]);
  const result = await found.executor.execute({
    operation: 'propose_memory',
    input: { ...explicitMemory(), sourceType: 'model_inferred', confidence: 0.9 },
  });

  assert.equal(result.status, 'input_required');
  assert.deepEqual(result.fields, ['confidence']);
  assert.deepEqual(found.calls, []);
});

test('list_memories validates and encodes paging filters', async () => {
  const found = executorFor([[OWNED_AGENT], { items: [], total: 0 }]);
  const result = await found.executor.execute({
    operation: 'list_memories', input: { limit: 25, offset: 5, status: 'pending_review' },
  });

  assert.equal(result.status, 'completed');
  assert.equal(
    found.calls[1].path,
    '/agent-network-api/agents/owned%2Fagent/memories?limit=25&offset=5&status=pending_review',
  );
});

test('delete_memory resolves ownership and uses a stable key', async () => {
  const found = executorFor([[OWNED_AGENT], { deleted: true }]);
  const result = await found.executor.execute({
    operation: 'delete_memory', input: { agentId: 'untrusted-agent', memoryId: 'memory/id' },
  });

  assert.equal(result.status, 'completed');
  assert.equal(found.calls[1].path, '/agent-network-api/agents/owned%2Fagent/memories/memory%2Fid');
  assert.equal(found.calls[1].idempotencyKey, 'workflow-1:memory:delete');
});

test('Node 与 Python 客户端记忆状态一致，且不接受平台来源', async () => {
  const [client, python] = await Promise.all([
    readFile(new URL('../src/memory-actions.js', import.meta.url), 'utf8'),
    readFile(new URL(
      '../skills/tokensmind-agent-network/scripts/memory_actions.py', import.meta.url,
    ), 'utf8'),
  ]);
  const statuses = ['active', 'deleted', 'pending_review', 'rejected'];
  for (const source of [client, python]) {
    for (const status of statuses) assert.ok(source.includes(`'${status}'`) || source.includes(`"${status}"`));
    // expired / superseded 没有任何写入路径，服务端已移除，客户端不能再发
    assert.doesNotMatch(source, /'expired'|"expired"|'superseded'|"superseded"/);
  }
  // platform_event 只能由服务端写入
  assert.doesNotMatch(client, /MEMORY_SOURCES[\s\S]{0,80}platform_event/);
  assert.doesNotMatch(python, /MEMORY_SOURCES[\s\S]{0,80}platform_event/);
});
